text/2020-03-11-versioned-kv.md

```diff
@@ -0,0 +1,196 @@
+# Versioned KV (VerKV)
+
+## Summary
+
+This proposal presents another KV interface - Versioned KV (VerKV), a
+multi-versioned KV between RawKV and TxnKV.
+
+## Motivation
+
+At present, Yidian Zixun adopts TiKV RawKV for Key-Value storage. Due to
+performance concerns, TxnKV is not used. However, RawKV only provides simple KV
+semantics, and does not support multiple versions based on timestamp,
+which causes some problems:
+
+- When services are migrated from HBase to TiKV, RawKV does not fulfil some
+  major functionality requirements such as data expiration and historical data
+   version;
+- When developing backup and synchronization functions based on RawKV,
+  it is difficult to solve the problems of online data migration, incremental
+   backup, and data synchronization between two data centers because timestamp
+  is not supported.
+
+In order to solve the above problems, this proposal presents another kv
+interface - VerKV, which mainly addsversion information such as timestamp to
+the key. At the same time, this interface will be designed to be approximate
+to the semantics of HBase, which will facilitate the service migration from
+HBase to TiKV, thereby integrating TiKV better into the HBase ecosystem.
+
+## Detailed designs
+
+### Functions and scenarios
+
+Here are the functionality designs and the corresponding application
+scenarios for VerKV:
+
+- Support data migration, incremental backup, CDC (Change Data Capture)
+  and cross-IDC cluster synchronization application scenarios include:
+
+  - Online data replication between TiKV clusters, which includes a full
+    migrations followed by incremental migrations;
+  - Cluster data backup, which includes a first full backup followed perform
+    an incremental data backups every other time;
+  - CDC service created based on incremental data acquisition; can reliably
+    provide incremental data services;
+  - In cross data center cluster synchronization scenarios, VerKV can not
+    only process incremental data synchronization, but also handle write
+    conflicts caused by the same key.
+
+- Support multi-versioned data. The corresponding application scenarios
+  include:
+
+  - Data version retrieval. For certain businesses, multiple versions of
+    historical data must be maintained for for backtracking and analysis.
+    Reliance on third-party storage and consumption of computing resources
+    (user dynamic characteristics, User behavior link tracking) needs to be
+    reduced, and the data processing process optimized;
+  - Data version control, for fast data recovery after problematic data
+    writes in businesses.
+
+- Support data expiration based on TTL. The corresponding application
+  scenarios include:
+
+  - Setting expiration time to obtain data of a recent period
+  - Setting expiration time to clear inactive data for a specified amount
+    of time
+
+### Compatibility
+
+The design adds a new Column Family called `VerCF` in RocksDB, which avoids
```

Comment:
maybe we can support multiple CF, by prefix all such CF with `VerCF_` prefix? Similarly we can prefix RawKV CFs with `RawCF_` prefix, then we can support mixed use of all these incompatible APIs.

Reply:
multple CF is not supported temporarily, but will reserve a field of cf for implementing multiple CF later.